### PR TITLE
refactor(event_handler): add type annotations for router decorators

### LINKED ISSUE #3990: Static typing: APIGatewayRestResolver `.post()` decorator return type Unknown
### Static type checker used

pyright/pylance

### AWS Lambda function runtime

3.12

### Powertools for AWS Lambda (Python) version

latest

### Static type checker info

Latest Pylance, latest VS Code, type checking mode strict.

```
Untyped function decorator obscures type of function
```

Screenshot:

image

### Code snippet

```python
from aws_lambda_powertools.event_handler import APIGatewayRestResolver

app = APIGatewayRestResolver()

@app.post("/v1/demo")
def demo() -> str:
    return "demo"
```

### Possible Solution

return `Callable[..., T]` or something like that

### PR DESCRIPTION
**Pull request by rafrafek**
**Issue number: #3990**

## Summary

Adds type annotations for Api Gateway Router decorators to improve Pyright compatibility. Fixes Pyright error `Untyped function decorator obscures type of function` when used in "strict" type checking mode. May be helpful especially for VS Code users.

### Changes

- Added type annotations for the return value of Api Gateway Router decorators.

### User experience

- Before the change the decorators were showing errors in VS Code when Pylance/Pyright was set to "strict" mode.
- Now there are no errors

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by rafrafek** [comment]
Should I use `AnyCallableT` from `aws_lambda_powertools/shared/types.py`?

File: aws_lambda_powertools/event_handler/api_gateway.py [hunk 1]
```diff
@@ -898,6 +898,7 @@ def build(self, event: ResponseEventT, cors: CORSConfig | None = None) -> dict[s
             **event.header_serializer().serialize(headers=self.response.headers, cookies=self.response.cookies),
         }
 
+T_route = TypeVar("T_route", bound=Callable[..., Any])
```
Comment (leandrodamascena) [inline aws_lambda_powertools/event_handler/api_gateway.py]: Please check if importing `AnyCallableT` from `aws_lambda_powertools/shared/types.py` works the same.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey @rafrafek! Thanks for sending this PR and sorry for delay in replying it.

Just a small comment about using an existing type instead of creating a new one. If it work, please import it under `TYPE_CHECKING` section.

File: aws_lambda_powertools/event_handler/api_gateway.py (hunk 1 above)
Comment (rafrafek) [reply aws_lambda_powertools/event_handler/api_gateway.py]: Hi @leandrodamascena, I did the import and it seems to be working!

**Review by leandrodamascena — APPROVED** [review]
Thanks a lot for addressing the feedback @rafrafek!

APPROVED!

**Pull request merged**